plugins/cc/ccprojectability.ml

```diff
@@ -0,0 +1,338 @@
+    | FromParameter (type_to_compose, type_to_compose_from, index, extraction)->
+      Pp.(
+        str level ++ str "compose: " ++ print env sigma type_to_compose ++ str "\n" ++
+        str level ++ str "from param: " ++ print env sigma type_to_compose_from ++ str "\n" ++
+        str level ++ str "at: " ++ int index ++ str "\n" ++
+        str level ++ str "by: " ++ _pr_type_extraction env sigma extraction ~level:(level ^"\t") ++ str "\n"
+      )
+    | FromIndex (type_to_compose, type_to_compose_from, index, extraction)->
+      Pp.(
+        str level ++ str "compose: " ++ print env sigma type_to_compose ++ str "\n" ++
+        str level ++ str "from index: " ++ print env sigma type_to_compose_from ++ str "\n" ++
+        str level ++ str "at: " ++ int index ++ str "\n" ++
+        str level ++ str "by:\n" ++ _pr_type_extraction env sigma extraction ~level:(level ^ "\t") ++ str "\n"
+      )
+    | Composition ((f,f_composition), args_compositions) ->
+      Pp.(
+        str level ++ str "compose: " ++ print env sigma f ++ str "(" ++ seq (List.map (fun (arg,_) -> print env sigma arg ++ str ", ") (Array.to_list args_compositions)) ++ str ")" ++ str "\n" ++
+                     pr_type_composition_rec f_composition (level ^ "\t") ++ str "\n" ++
+        str level ++ str "(" ++ str "\n" ++
+                     seq (List.map (fun (_,c) -> pr_type_composition_rec c (level ^ "\t") ++ str ", ") (Array.to_list args_compositions)) ++ str "\n" ++
+        str level ++ str ")" ++ str "\n"
+      )
+  in pr_type_composition_rec type_composition level
+
+type projection_type =
+| Simple
+| Dependent of type_composition
+| NotProjectable
+
+(*find a composition to form a given term by extracting from given terms*)
+let find_type_composition
+  env sigma (cnstr:Names.constructor * EConstr.EInstance.t) (argindex:int) (env_field_type:EConstr.t) (ind:Names.inductive EConstr.puniverses) (env_ind_params:EConstr.t array) (env_ind_args:EConstr.t array)
+  : (Evd.evar_map * type_composition option) =
+  let env_ind_n_params = Array.length env_ind_params in
+  let rel_type_of_constructor = Inductiveops.type_of_constructor env cnstr in
+  let (rel_field_context, (rel_field_annot, rel_field_type), rel_field_rest) = get_ith_arg sigma (argindex+env_ind_n_params) rel_type_of_constructor in
+  let (rel_target_context, rel_target_type) = EConstr.decompose_prod sigma rel_field_rest in
+  let rel_field_type_lifted = EConstr.Vars.lift (List.length rel_target_context + 1) rel_field_type in
+  let (_,rel_args) = EConstr.decompose_app sigma rel_target_type in
+  let (rel_ind_params, rel_ind_args) = CArray.chop env_ind_n_params rel_args in
+  let rec find_type_composition_rec env sigma (rel_term_to_compose:EConstr.t) (env_term_to_compose:EConstr.t)=
+    let rel_term_to_compose_kind = EConstr.kind sigma rel_term_to_compose in
+    match rel_term_to_compose_kind with
+    | Var _ | Const _ | Ind _ | Construct _ -> (sigma, Some (FromEnv rel_term_to_compose))
+    | _ ->(
```

Comment:
You don't need to wrap the contents of a pattern-matching branch in parentheses (and neither of if-then-else). In case you need nested matches (and only when needed), to work around the deficient OCaml syntax I'd recommend wrapping the inner match in a begin-end block, like:
```ocaml
match foo with
| Some x ->
  begin match bar with
  | None -> ...
  | Some y -> ...
  end
| None -> ...
```